File: ios/Sources/ReleasebirdPlugin/ReleasebirdPlugin.swift
import Foundation
import Capacitor
/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(ReleasebirdPlugin)
public class ReleasebirdPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ReleasebirdPlugin"
    public let jsName = "Releasebird"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "echo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "showWidget", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "identify", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "logout", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = ReleasebirdCapacitor()

    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }
    
    @objc func initialize(_ call: CAPPluginCall) {
    DispatchQueue.global(qos: .userInitiated).async {
     DispatchQueue.main.async { [self] in
        let apiKey = call.getString("apiKey") ?? ""
        let showButton = call.getBool("showButton") ?? true
        implementation.initialize(apiKey, showButton: showButton);
        }
        }
        call.resolve();
    }
    
    @objc func showWidget(_ call: CAPPluginCall) {
    DispatchQueue.global(qos: .userInitiated).async {
     DispatchQueue.main.async { [self] in
        implementation.showWidget()
        }

    }
    call.resolve()
    }

    @objc func logout(_ call: CAPPluginCall) {
        DispatchQueue.global(qos: .userInitiated).async {
         DispatchQueue.main.async { [self] in
            implementation.logout()
            }

        }
        call.resolve()
    }
    
    func convertJSObjectToNSDictionary(jsObject: JSObject?) -> NSDictionary? {
        guard let jsObject = jsObject else {
            return nil
        }
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: jsObject, options: [])
            let dictionary = try JSONSerialization.jsonObject(with: jsonData, options: .mutableContainers) as? NSDictionary
            return dictionary
        } catch {
            print("Error converting JSObject: \(error.localizedDescription)")
            return nil
        }
    }
    
    @objc func identify(_ call: CAPPluginCall) {
        guard let identifyJson = call.getObject("identify") else {
            call.reject("identify is missing")
            return
        }
        guard let dictionary = convertJSObjectToNSDictionary(jsObject: identifyJson) else {
            call.reject("Failed to convert JSObject to NSDictionary")
            return
        }
        let hash = call.getString("hash") ?? ""
        implementation.identify(dictionary, andHash: hash)
        call.resolve()
    }

}

File: android/src/main/java/com/releasebird/capacitor/ReleasebirdPlugin.java
package com.releasebird.capacitor;

import android.util.Log;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

@CapacitorPlugin(name = "Releasebird")
public class ReleasebirdPlugin extends Plugin {

    private Releasebird implementation = new Releasebird();

    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");

        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }


    @PluginMethod
    public void initialize(PluginCall call) {
        String apiKey = call.getString("apiKey");
        implementation.initialize(apiKey, getActivity().getApplication());
        call.resolve();
    }

    @PluginMethod
    public void showWidget(PluginCall call) {
        implementation.showWidget();
        call.resolve();
    }

    @PluginMethod
    public void logout(PluginCall call) {
        implementation.logout(getActivity().getApplication());
        call.resolve();
    }

    @PluginMethod
    public void identify(PluginCall call) {
            String hash = call.getString("hash");

            JSObject identifyJson = call.getObject("identify");

            HashMap<String, Object> hashMap = new HashMap<>();

            // Get the iterator from the keys
            Iterator<String> keysIterator = identifyJson.keys();

            // Convert Iterator to a Set to use foreach loop
            Set<String> keysSet = new HashSet<>();
            while (keysIterator.hasNext()) {
                keysSet.add(keysIterator.next());
            }

            // Iterate through the keys using foreach
            for (String key : keysSet) {
                try {
                    Object value = identifyJson.get(key);
                    hashMap.put(key, value);
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
            implementation.identify(hashMap, hash);
            call.resolve();

    }
}
